File: esio/x-plane.f90
!!-----------------------------------------------------------------------bl-
!!--------------------------------------------------------------------------
!!
!! esio 0.1.6: ExaScale IO library for turbulence simulation restart files
!! http://pecos.ices.utexas.edu/
!!
!! Copyright (C) 2010 The PECOS Development Team
!!
!! This library is free software; you can redistribute it and/or
!! modify it under the terms of the Version 2.1 GNU Lesser General
!! Public License as published by the Free Software Foundation.
!!
!! This library is distributed in the hope that it will be useful,
!! but WITHOUT ANY WARRANTY; without even the implied warranty of
!! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
!! Lesser General Public License for more details.
!!
!! You should have received a copy of the GNU Lesser General Public
!! License along with this library; if not, write to the Free Software
!! Foundation, Inc. 51 Franklin Street, Fifth Floor,
!! Boston, MA  02110-1301  USA
!!
!!-----------------------------------------------------------------------el-
!! $Id$

! Designed to be #included from esio.f90 within a subroutine declaration

#if !defined(FINTENT) || !defined(CTYPE) || !defined(CBINDNAME)
#error "One of FINTENT, CTYPE, or CBINDNAME not defined"
#endif

  type(esio_handle), intent(in)            :: handle
  character(len=*),  intent(in)            :: name
#ifndef VECTORVALUED
  CTYPE,             FINTENT               :: plane(1,*)
#else
  CTYPE,             FINTENT               :: plane(1,1,*)
  integer,           intent(in)            :: ncomponents
#endif
  integer,           intent(in),  optional :: astride
  integer,           intent(in),  optional :: bstride
#ifdef HASCOMMENT
  character(len=*),  intent(in),  optional :: comment
#endif
  integer,           intent(out), optional :: ierr

  integer :: stat, tmp_astride, tmp_bstride
#ifdef HASCOMMENT
  character(len=1,kind=c_char), parameter :: null_char(1) = (/ c_null_char /)
#endif

#ifndef DOXYGEN_SHOULD_SKIP_THIS
  interface
    function plane_impl (handle, name, plane,        &
                         bstride, astride            &
#ifdef VECTORVALUED
                         ,ncomponents                &
#endif
#ifdef HASCOMMENT
                        ,comment                     &
#endif
                        ) bind (C, name=CBINDNAME)
      import
      integer(c_int)                                  :: plane_impl
      type(esio_handle),            intent(in), value :: handle
      character(len=1,kind=c_char), intent(in)        :: name(*)
      CTYPE,                        FINTENT           :: plane(*)
      integer(c_int),               intent(in), value :: bstride
      integer(c_int),               intent(in), value :: astride
#ifdef VECTORVALUED
      integer(c_int),               intent(in), value :: ncomponents
#endif
#ifdef HASCOMMENT
      character(len=1,kind=c_char), intent(in)        :: comment(*)
#endif
    end function plane_impl
  end interface
#endif /* DOXYGEN_SHOULD_SKIP_THIS */

  tmp_astride = 0
  if (present(astride)) tmp_astride = astride
  tmp_bstride = 0
  if (present(bstride)) tmp_bstride = bstride

! Note reordering Fortran's (a, b) to C's (b, a)
#ifndef HASCOMMENT
  stat = plane_impl(handle, esio_f_c_string(name), plane, &
                    tmp_bstride, tmp_astride              &
#ifdef VECTORVALUED
                    ,ncomponents                          &
#endif
                   )
#else /* HASCOMMENT */
  if (present(comment)) then
    stat = plane_impl(handle, esio_f_c_string(name), plane, &
                      tmp_bstride, tmp_astride              &
#ifdef VECTORVALUED
                      ,ncomponents                          &
#endif
                     ,esio_f_c_string(comment)              &
                    )
  else
    stat = plane_impl(handle, esio_f_c_string(name), plane, &
                      tmp_bstride, tmp_astride              &
#ifdef VECTORVALUED
                      ,ncomponents                          &
#endif
                     ,null_char                             &
                    )
  end if
#endif /* HASCOMMENT */

  if (present(ierr)) ierr = stat

#undef FINTENT
#undef CTYPE
#undef CBINDNAME
#ifdef HASCOMMENT
#undef HASCOMMENT
#endif
#ifdef VECTORVALUED
#undef VECTORVALUED
#endif

File: esio/x-plane.c
//-----------------------------------------------------------------------bl-
//--------------------------------------------------------------------------
//
// esio 0.1.6: ExaScale IO library for turbulence simulation restart files
// http://pecos.ices.utexas.edu/
//
// Copyright (C) 2010 The PECOS Development Team
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the Version 2.1 GNU Lesser General
// Public License as published by the Free Software Foundation.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc. 51 Franklin Street, Fifth Floor,
// Boston, MA  02110-1301  USA
//
//-----------------------------------------------------------------------el-
// $Id$

// Designed to be #included from layout.c
#if !defined(METHODNAME) || !defined(OPFUNC) || !defined(QUALIFIER)
#error "One of METHODNAME, OPFUNC, or QUALIFIER not defined"
#endif

hid_t METHODNAME(hid_t plist_id, hid_t dset_id, QUALIFIER void *plane,
                 int bglobal, int bstart, int blocal, int bstride,
                 int aglobal, int astart, int alocal, int astride,
                 hid_t type_id)
{
    (void) bglobal; /* Unused but present for API consistency */
    (void) aglobal; /* Unused but present for API consistency */

    /* Establish (possibly strided) memspace details */
    const hsize_t nelems = blocal * bstride;
    const hsize_t lies   = 1;
    const hid_t memspace
        = H5Screate_simple(1, blocal * alocal ? &nelems : &lies, NULL);
    assert(memspace > 0);
    if (blocal * alocal == 0) {
        H5Sselect_none(memspace);
    } else if (astride != 1 || bstride != astride * alocal) {
        /* Strided memspace; additional hyperslab selection necessary */
        if (H5Sselect_none(memspace) < 0) {
            H5Sclose(memspace);
            ESIO_ERROR("Resetting memory hyperslab failed", ESIO_EFAILED);
        }
        for (int j = 0; j < blocal; ++j) {
            const hsize_t start  = j*bstride;
            const hsize_t stride = astride;
            const hsize_t count  = alocal;
            if (H5Sselect_hyperslab(memspace, H5S_SELECT_OR,
                                    &start, &stride, &count, NULL) < 0) {
                H5Sclose(memspace);
                ESIO_ERROR("Selecting memory hyperslab failed", ESIO_EFAILED);
            }
        }
    }

    /* Establish (contiguous) filespace details */
    const hid_t filespace = H5Dget_space(dset_id);
    assert(filespace >= 0);
    const hsize_t start[2] = { bstart, astart };
    const hsize_t count[2] = { blocal, alocal };
    if (blocal * alocal == 0) {
        H5Sselect_none(filespace);
    } else if (H5Sselect_hyperslab(filespace, H5S_SELECT_SET,
                                   start, NULL, count, NULL) < 0) {
        H5Sclose(memspace);
        H5Sclose(filespace);
        ESIO_ERROR("Selecting file hyperslab failed", ESIO_EFAILED);
    }

    /* Transfer hyperslab to or from memory */
    const herr_t status = OPFUNC(dset_id, type_id, memspace,
                                 filespace, plist_id, plane);
    if (status < 0) {
        H5Sclose(filespace);
        H5Sclose(memspace);
        ESIO_ERROR("Operation failed", ESIO_EFAILED);
    }

    /* Release temporary resources */
    H5Sclose(filespace);
    H5Sclose(memspace);

    return ESIO_SUCCESS;
}
